core/src/main/java/io/stargate/core/activator/BaseActivator.java

```diff
@@ -0,0 +1,239 @@
+   * notification denoting that service was registered using {@link
+   * Tracker#addingService(ServiceReference)}. If all services are present, it will call the
+   * user-provided {@link this#createService(List)} and register it in the OSGi using {@link
+   * BundleContext#registerService(Class, Object, java.util.Dictionary)} if {@code
+   * targetServiceClass.isPresent()}.
+   */
+  @Override
+  public synchronized void start(BundleContext context) throws InvalidSyntaxException {
+    logger.info("Starting {} ...", activatorName);
+    this.context = context;
+    tracker =
+        new Tracker(
+            context, context.createFilter(constructDependenciesFilter()), registeredServices);
+    tracker.open();
+  }
+
+  @VisibleForTesting
+  String constructDependenciesFilter() {
+    StringBuilder builder = new StringBuilder("(|");
+
+    for (ServiceDependency serviceDependency : serviceDependencies) {
+      if (serviceDependency.identifier.isPresent()) {
+        builder.append(String.format("(%s)", serviceDependency.identifier.get()));
+      } else {
+        builder.append(String.format("(objectClass=%s)", serviceDependency.className.getName()));
+      }
+    }
+    builder.append(")");
+    return builder.toString();
+  }
+
+  @Override
+  public synchronized void stop(BundleContext context) throws Exception {
+    if (started) {
+      logger.info("Stopping {}", activatorName);
+      stopService();
+    }
+    tracker.close();
+  }
+
+  private synchronized void startServiceInternal(List<Object> dependentServices) {
+    if (started) {
+      logger.info("The {} is already started. Ignoring the start request.", activatorName);
+      return;
+    }
+    started = true;
+    ServiceAndProperties service = createService(dependentServices);
+    if (service != null && targetServiceClass.isPresent()) {
+      logger.info("Registering {} as {}", activatorName, targetServiceClass.get().getName());
+      context.registerService(
+          targetServiceClass.get().getName(), service.service, service.properties);
+    }
+    logger.info("Started {}", activatorName);
+  }
+
+  public class Tracker extends ServiceTracker<Object, Object> {
+
+    private final LinkedHashMap<Class<?>, Object> registeredServices;
+
+    public Tracker(
+        BundleContext context, Filter filter, LinkedHashMap<Class<?>, Object> registeredServices) {
+      super(context, filter, null);
+      this.registeredServices = registeredServices;
+    }
+
+    /**
+     * It will try to match all dependentServices with a ServiceReference notification. After the
+     * notification is handled, it checks if all dependentServices are not null. If they are not,
+     * the client's provided {@link this#createService(List)} is called, and the service is
+     * registered. It will not register the service if it was already registered.
+     */
+    @Override
+    public Object addingService(ServiceReference<Object> ref) {
```

Comment:
Should we also handle service removal? I'd guess if a dependency disappears we should notify the service impl. sub-class. In the simple case I suppose it makes sense to stop the service. When the dependency re-appears the service will be restarted.

Services that need more complex dependency tracking logic will probably have to use custom logic (i.e. not extend `BaseActivator`)... WDYT?

Reply:
From what I see, the use case that you described (restart service) is not implemented in any existing `Activator`.
I am ok with having that, but I think this should be a follow-up ticket. I would like this PR to have only refactoring changes without adding new features.

Reply:
sgtm